packages/react-slider/src/components/Slider/useSliderState.tsx

```diff
@@ -0,0 +1,238 @@
+      const distance = thumbPosition - position;
+      return distance / stepLength;
+    },
+    [max, min, step],
+  );
+
+  const onPointerMove = React.useCallback(
+    (ev: React.PointerEvent<HTMLDivElement>): void => {
+      if (step !== 1) {
+        updateValue(Math.round((min + step * calculateSteps(ev)) / step) * step, ev);
+      } else {
+        updateValue(min + step * calculateSteps(ev), ev);
+      }
+    },
+    [calculateSteps, min, step, updateValue],
+  );
+
+  const onPointerUp = (): void => {
+    disposables.current.forEach(dispose => dispose());
+    disposables.current = [];
+    thumbRef.current!.focus();
+  };
+
+  const onPointerDown = React.useCallback(
+    (ev: React.PointerEvent<HTMLDivElement>): void => {
+      const { pointerId } = ev;
+      const target = ev.target as HTMLElement;
+
+      if (target.setPointerCapture) {
+        target.setPointerCapture(pointerId);
+      }
+
+      onPointerDownCallback?.(ev);
+
+      disposables.current.push(on(target, 'pointermove', onPointerMove), on(target, 'pointerup', onPointerUp), () => {
+        target.releasePointerCapture(pointerId);
+      });
+
+      onPointerMove(ev);
+    },
+    [onPointerMove, onPointerDownCallback],
+  );
+
+  const onKeyDown = React.useCallback(
+    (ev: React.KeyboardEvent<HTMLDivElement>): void => {
+      onKeyDownCallback?.(ev);
+
+      if (ev.shiftKey) {
+        if (ev.key === 'ArrowDown' || ev.key === 'ArrowLeft') {
+          updateValue(currentValue! - step * 10, ev);
+          return;
+        } else if (ev.key === 'ArrowUp' || ev.key === 'ArrowRight') {
+          updateValue(currentValue! + step * 10, ev);
+          return;
+        }
+      } else if (ev.key === 'ArrowDown' || ev.key === 'ArrowLeft') {
+        updateValue(currentValue! - step, ev);
+        return;
+      } else if (ev.key === 'ArrowUp' || ev.key === 'ArrowRight') {
+        updateValue(currentValue! + step, ev);
+        return;
+      } else {
+        switch (ev.key) {
+          case 'PageDown':
+            updateValue(currentValue! - step * 10, ev);
+            break;
+          case 'PageUp':
+            updateValue(currentValue! + step * 10, ev);
+            break;
+          case 'Home':
+            updateValue(min, ev);
+            break;
+          case 'End':
+            updateValue(max, ev);
+            break;
+        }
+      }
+    },
+    [currentValue, max, min, onKeyDownCallback, step, updateValue],
+  );
+
+  React.useEffect(() => {
+    if (state.ref && state.ref.current) {
+      state.ref.current.value = currentValue;
+      state.ref.current.focus = () => thumbRef?.current?.focus();
+    }
+  }, [currentValue, state.ref]);
+
+  useMount(() => {
+    // If the user passes an out of bounds controlled value, clamp and update their value onMount.
+    if (value !== undefined && (value < min || value > max) && onChange && onChangeCallback.current) {
+      onChangeCallback.current(clamp(value, min, max));
```

Comment:
I don't think `onChange` should be called here (we don't call it on clamping today)

Reply:
That's fair, I'll look into removing the onChange here